### ISSUE TITLE
Implement a review backend

### ISSUE THREAD

**Issue by razzeee** [flathub/website#50]

**Comment by razzeee** [comment]
Should wait for https://gitlab.gnome.org/Infrastructure/odrs-web/-/merge_requests/17 so that we can filter for actual reviews of the flatpak versions

**Comment by sonnyp** [comment]
odrs lack features that are important for publishers, such as replying to reviews publicly or privately contacting a review author to ask for more details.

Would these be doable wth odrs?

**Comment by razzeee** [comment]
I'm wondering if just starting with stars would be any improvement.

**Comment by mijorus** [comment]
Hi @razzeee, I would like to contribute to this feature.
  
Is this https://github.com/flathub-infra/website/issues/50#issuecomment-1179700245 this a blocker for you?
I know that mixing reviews from multiple sources is not ideal, but as a user and an app developer it would be great to see all my reviews on a single place.

I basically had to make my [own implementation](https://flatstat.mijorus.it/app/it.mijorus.smile) of that, just to see the reviews of my app

**Comment by razzeee** [comment]
hey,

> Is this https://github.com/flathub-infra/website/issues/50#issuecomment-1179700245 this a blocker for you?

I at least thought so, as I suspect reviews heavily diff between the way something is packaged. 

> I know that mixing reviews from multiple sources is not ideal, but as a user and an app developer it would be great to see all my reviews on a single place.

Not sure why as a `user` - I do get that as a `developer`, but thought you could also just login to odrs for that (I've only patched that thing a few times, never really used it)

**Comment by mijorus** [comment]
I think it makes sense to show reviews and star ratings at least for verified apps, as the developer is actively targeting Flathub, so there shouldn't be issues related to sandboxing.


>  but thought you could also just login to odrs for that 

Can I? Not really sure where or how, their website says I can "request an account if you think it is required", without clarifying where to ask for.

**Comment by mijorus** [comment]
I'll send a pr anyway, then you can decide what you think about it

**Comment by razzeee** [comment]
Please wirte an outline of what you are planning to do first

**Comment by mijorus** [comment]
Sure. This sis my proposal

1. We start by implementing the rating system first, comments and reviews later, as we need to decide how to handle multiple languages
2. Only verified apps should have visible reviews: since the developer is officially publishing the Flathub release, I think we can assume that the app should work flawlessly and is being developed with Flatpak's limitations in mind
3. We can create a backend api that caches ratings from ODRS, say once a day, with redis: the API responds with the average rating already computed
4. (Optional) We can also computer our own average, by filtering out all the reviews published before the flathub release, but we need to verify if ODRS apis can handle such queries.

**Comment by razzeee** [comment]
Sounds good. Just trying to figure out if your proposal would conflict with this https://github.com/flathub-infra/gsoc#flathub-rating-system

**Comment by mijorus** [comment]
> Sounds good. Just trying to figure out if your proposal would conflict with this https://github.com/flathub-infra/gsoc#flathub-rating-system

Well, ODRS apis are definitely a bit rough. They don't support timestamps queries, search or pagination, which means they return the full payload everytime, which is not great.

Also, `/ratings/<app_id>` doesn't work, which means you need to download the full list just to get ratings for one app. It was not really designed with scale in mind.

**Comment by razzeee** [comment]
you don't have to explain that :)

but we should probably keep those endpoints around, if we manage to redo it, breaking every software center app is likely a nogo

**Comment by mijorus** [comment]
> you don't have to explain that :)
> 
> but we should probably keep those endpoints around, if we manage to redo it, breaking every software center app is likely a nogo

So, do I have green light on this one? Can I open a PR with a sample implementation? Let me know what you think.
(Like I said before, I would start by implementing the rating system first, aka the stars)

**Comment by razzeee** [comment]
Yeah, as that would still rely on having odrs to consume it, that would not make the gsoc proposal invalid. (making it invalid would also be fine, but then I should remove that pitch)

**Comment by mijorus** [comment]
HI @razzeee , I'm having an hard time figuring out where to refresh app ratings.
I tough it could be a good idea to queue the refresh to any cronjob worker you might already have.

I came across `backend/app/worker.py`, but I'm not really sure 100% if this is the right place

**Comment by razzeee** [comment]
It is invoked multiple times a day, but at this point, we could probably just have a once per day job and we will figure out it being called once per day on the server. So feel free to create such a thing.

**Comment by mijorus** [comment]
> It is invoked multiple times a day, but at this point, we could probably just have a once per day job and we will figure out it being called once per day on the server. So feel free to create such a thing.

All right, I guess it should always be an http endpoint or can we run python scripts directly?

**Comment by barthalion** [comment]
https://dramatiq.io/cookbook.html#scheduling-messages

**Comment by razzeee** [comment]
IMO this is just not happening at all. Nobody maintains ODRS in a decent way. We could work it into our side of things, but that's only helpful for us and will also mean, we need to bear the moderation burden.
